Plutarch/Api/V1/Scripts.hs

```diff
@@ -64,3 +113,230 @@ deriving via
+
+-- | a Validator Term
+type PValidator = PData :--> PData :--> PData :--> POpaque
+
+-- | a MintingPolicy Term
+type PMintingPolicy = PData :--> PData :--> POpaque
+
+-- | a StakeValidator Term
+type PStakeValidator = PData :--> PData :--> POpaque
+
+-- | A compiled Validator, bundled with the Plutarch Term
+data Validator = Validator
+  { validatorScript :: Plutus.Script
+  , validatorTerm :: ClosedTerm PValidator
+  }
+
+-- | A compiled MintingPolicy, bundled with the Plutarch Term
+data MintingPolicy = MintingPolicy
+  { mintingPolicyScript :: Plutus.Script
+  , mintingPolicyTerm :: ClosedTerm PMintingPolicy
+  }
+
+-- | A compiled StakeValidator, bundled with the Plutarch Term
+data StakeValidator = StakeValidator
+  { stakeValidatorScript :: Plutus.Script
+  , stakeValidatorTerm :: ClosedTerm PStakeValidator
+  }
+
+-- | Compile a Validator
+mkValidator :: ClosedTerm PValidator -> Validator
+mkValidator term =
+  Validator (compile term) term
+
+-- | Compile a MintingPolicy
+mkMintingPolicy :: ClosedTerm PMintingPolicy -> MintingPolicy
+mkMintingPolicy term =
+  MintingPolicy (compile term) term
+
+-- | Compile a StakeValidator
+mkStakeValidator :: ClosedTerm PStakeValidator -> StakeValidator
+mkStakeValidator term =
+  StakeValidator (compile term) term
+
+-- | Serialise a Validator to a (Lazy) ByteString
+serialiseValidator :: Validator -> Lazy.ByteString
+serialiseValidator = serialiseScript . validatorScript
+
+-- | Serialise a StakeValidator to a (Lazy) ByteString
+serialiseStakeValidator :: StakeValidator -> Lazy.ByteString
+serialiseStakeValidator = serialiseScript . stakeValidatorScript
+
+-- | Serialise a MintingPolicy to a (Lazy) ByteString
+serialiseMintingPolicy :: MintingPolicy -> Lazy.ByteString
+serialiseMintingPolicy = serialiseScript . mintingPolicyScript
+
+-- | Hash a Script, with the correct prefix for Plutus V1
+scriptHash :: Plutus.Script -> Plutus.ScriptHash
+scriptHash = hashScriptWithPrefix "\x01"
+
+-- | Hash a Validator, with the correct prefix for Plutus V1
+validatorHash :: Validator -> Plutus.ValidatorHash
+validatorHash = coerce . scriptHash . validatorScript
+
+-- | Hash a MintingPolicy, with the correct prefix for Plutus V1
+mintingPolicySymbol :: MintingPolicy -> Plutus.CurrencySymbol
+mintingPolicySymbol = coerce . scriptHash . mintingPolicyScript
+
+-- | Hash a StakeValidator, with the correct prefix for Plutus V1
+stakeValidatorHash :: StakeValidator -> Plutus.StakeValidatorHash
+stakeValidatorHash = coerce . scriptHash . stakeValidatorScript
+
+{- |
+  Construct a `PValidator`, assuming the `PData` arguments are
+  of the correct form.
+
+  Shorthand for:
+  @pwrapValidatorWith punsafeAsData punsafeAsData punsafeAsData@
+-}
+pwrapValidatorAsData ::
+  ( forall s.
+    Term s (PAsData datum) ->
+    Term s (PAsData redeemer) ->
+    Term s (PAsData ctx) ->
+    Term s p
+  ) ->
+  ClosedTerm PValidator
+pwrapValidatorAsData = pwrapValidatorWith punsafeAsData punsafeAsData punsafeAsData
```

Comment:
there are a good amount of cases where you _would_ want to just assume the shape of Data, without checking it. 
Though I agree doing so should be more explicit, and not encouraged as a default...
I don't know if I like `PAsDataMaybe` as a wrapper - wouldn't we then need to (transitively!) wrap everything 'downstream' in a 'maybe', similarly?

Reply:
Probably, but it's usually not going to be a deeply nested type. A possible vulnerability that might not be obvious to users, is setting a global UTXO that contains some funds, the datum of which you can change by supplying a redeemer that is a list of integers. The new datum is the old datum, with the list of integers appended to the old list of integers. The issue is that with this, users would do `punsafeFromData` or use whatever abstraction you have on top of it, and assume as with PlutusTx that it in fact validates the data, when it does not. This would result in a UTXO that contains the global funds, with an invalid datum. This might make it unspendable, and thus be a vector for an attack.

Reply:
Yeah, I agree on the worry there. I removed the unsafe functions.
It's probably best left up to users to write explicit checks as necessary for their Datum/Redeemer types.